types/wlr_virtual_keyboard.c

```diff
@@ -0,0 +1,209 @@
+		goto context_fail;
+	}
+	void *data = mmap(NULL, size, PROT_READ, MAP_PRIVATE, fd, 0);
+	if (!data) {
+		goto fd_fail;
+	}
+	struct xkb_keymap *keymap = xkb_keymap_new_from_string(context, data,
+		XKB_KEYMAP_FORMAT_TEXT_V1, XKB_KEYMAP_COMPILE_NO_FLAGS);
+	if (!keymap) {
+		goto keymap_fail;
+	}
+	wlr_keyboard_set_keymap(keyboard->input_device.keyboard, keymap);
+	xkb_keymap_unref(keymap);
+	xkb_context_unref(context);
+	return;
+keymap_fail:
+	munmap(data, size);
+fd_fail:
+	xkb_context_unref(context);
+context_fail:
+	wl_client_post_no_memory(client);
+	return;
+}
+
+static void virtual_keyboard_key(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t key,
+		uint32_t state) {
+	struct wlr_virtual_keyboard *keyboard =
+		virtual_keyboard_from_resource(resource);
+	struct wlr_event_keyboard_key event = {
+		.time_msec = time,
+		.keycode = key,
+		.update_state = false,
+		.state = state,
+	};
+	wlr_keyboard_notify_key(keyboard->input_device.keyboard, &event);
+}
+
+static void virtual_keyboard_modifiers(struct wl_client *client,
+		struct wl_resource *resource, uint32_t mods_depressed,
+		uint32_t mods_latched, uint32_t mods_locked, uint32_t group) {
+	struct wlr_virtual_keyboard *keyboard =
+		virtual_keyboard_from_resource(resource);
+	wlr_keyboard_notify_modifiers(keyboard->input_device.keyboard,
+		mods_depressed, mods_latched, mods_locked, group);
+}
+
+static void virtual_keyboard_destroy_resource(struct wl_resource *resource) {
+	struct wlr_virtual_keyboard *keyboard =
+		virtual_keyboard_from_resource(resource);
+	wlr_signal_emit_safe(&keyboard->events.destroy, keyboard);
+	wlr_keyboard_destroy(keyboard->input_device.keyboard);
+	free(keyboard);
+}
+
+static void virtual_keyboard_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	virtual_keyboard_destroy_resource(resource);
+}
+
+static const struct zwp_virtual_keyboard_v1_interface virtual_keyboard_impl = {
+	.keymap = virtual_keyboard_keymap,
+	.key = virtual_keyboard_key,
+	.modifiers = virtual_keyboard_modifiers,
+	.destroy = virtual_keyboard_destroy,
+};
+
+static const struct zwp_virtual_keyboard_manager_v1_interface manager_impl;
+
+static struct wlr_virtual_keyboard_manager *manager_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwp_virtual_keyboard_manager_v1_interface, &manager_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void virtual_keyboard_manager_create_virtual_keyboard(
+		struct wl_client *client, struct wl_resource *resource,
+		struct wl_resource *seat,
+		uint32_t id) {
+	struct wlr_virtual_keyboard_manager *manager =
+		manager_from_resource(resource);
+
+	struct wlr_virtual_keyboard *virtual_keyboard = calloc(1,
+		sizeof(struct wlr_virtual_keyboard));
+	if (!virtual_keyboard) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wlr_keyboard* keyboard = calloc(1, sizeof(struct wlr_keyboard));
+	if (keyboard == NULL) {
```

Comment:
Should be

if (!keyboard) {

for consistency.

Reply:
Nah, both are fine

Reply:
The style is mixed in the same function though... your call.